hw/arm/guest-socket.c

```diff
@@ -49,44 +117,59 @@ int32_t qc_handle_socket(CPUState *cpu, int32_t domain, int32_t type,
         guest_svcs_errno = ENOTSOCK;
     } else if ((guest_svcs_fds[retval] = socket(domain, type, protocol)) < 0) {
         retval = -1;
-        guest_svcs_errno = errno;
+        guest_svcs_errno = convert_error(errno);
     }
 
     return retval;
 }
 
-int32_t qc_handle_accept(CPUState *cpu, int32_t sckt, struct sockaddr *g_addr,
-                         socklen_t *g_addrlen)
+int32_t qc_handle_accept(CPUState *cpu, int32_t sckt, S_SOCKADDR *g_addr,
+                         SOCKLEN_T *g_addrlen)
 {
     struct sockaddr_in addr;
-    socklen_t addrlen;
+    SOCKLEN_T addrlen;
 
+    addrlen = sizeof(addr);
     VERIFY_FD(sckt);
 
     int retval = find_free_socket();
 
     // TODO: timeout
     if (retval < 0) {
         guest_svcs_errno = ENOTSOCK;
-    } else if ((guest_svcs_fds[retval] = accept(guest_svcs_fds[sckt],
+    } else if ((guest_svcs_fds[retval] =
+#ifdef NETWORK_HACK
+                                         accept4(guest_svcs_fds[sckt],
+                                         (struct sockaddr *) &addr,
+                                         &addrlen, SOCK_NONBLOCK)) < 0)
+#else
+                                         accept(guest_svcs_fds[sckt],
                                          (struct sockaddr *) &addr,
-                                         &addrlen)) < 0) {
+                                         &addrlen)) < 0)
+#endif
+    {
         retval = -1;
-        guest_svcs_errno = errno;
+        guest_svcs_errno = convert_error(errno);
     } else {
-        cpu_memory_rw_debug(cpu, (target_ulong) g_addr, (uint8_t*) &addr,
-                            sizeof(addr), 1);
+#ifdef NETWORK_HACK
+        S_SOCKADDR_IN ADDR = convert_sockaddr_from(addr);
+#endif
+        cpu_memory_rw_debug(cpu, (target_ulong) g_addr, (uint8_t*) &ADDR,
+                            sizeof(ADDR), 1);
         cpu_memory_rw_debug(cpu, (target_ulong) g_addrlen,
                             (uint8_t*) &addrlen, sizeof(addrlen), 1);
     }
 
     return retval;
 }
 
-int32_t qc_handle_bind(CPUState *cpu, int32_t sckt, struct sockaddr *g_addr,
-                       socklen_t addrlen)
+int32_t qc_handle_bind(CPUState *cpu, int32_t sckt, S_SOCKADDR *g_addr,
+                       SOCKLEN_T addrlen)
 {
     struct sockaddr_in addr;
+#ifdef NETWORK_HACK
+    S_SOCKADDR_IN ADDR;
+#endif
```

Comment:
This will necessarily be `darwin_sockaddr_in darwin_addr`, if `__APPLE__` is defined...